packages/@aws-cdk/aws-codepipeline/lib/pipeline.ts

```diff
@@ -405,52 +434,32 @@ export class Pipeline extends PipelineBase {
   }
 
   private ensureReplicationResourcesExistFor(action: IAction): CrossRegionInfo {
-    const pipelineStack = Stack.of(this);
+    const actionEnv = new EnhancedAction(action).env;
 
-    let actionRegion: string | undefined;
-    let otherStack: Stack | undefined;
+    // Whether the action seems to care about its account/region (or is happy
+    // to inherit from the pipeline)
+    const hasAccount = !Token.isUnresolved(actionEnv.account);
+    const hasRegion = !Token.isUnresolved(actionEnv.region);
 
-    const actionResource = action.actionProperties.resource;
-    if (actionResource) {
-      const pipelineAndActionRegionComparison = Token.compareStrings(this.env.region, actionResource.env.region);
-      const pipelineAndActionInDifferentRegions = pipelineAndActionRegionComparison === TokenComparison.ONE_UNRESOLVED ||
-        pipelineAndActionRegionComparison === TokenComparison.DIFFERENT;
-      if (pipelineAndActionInDifferentRegions) {
-        actionRegion = actionResource.env.region;
-
-        // if the resource is from a different stack in another region but the same account,
-        // use that stack as home for the cross-region support resources
-        const actionResourceStack = Stack.of(actionResource);
-        const actionResourceAndItsStackRegionComparison = Token.compareStrings(actionResource.env.region, actionResourceStack.region);
-        const actionResourceInSameRegionAsItsStack = actionResourceAndItsStackRegionComparison === TokenComparison.SAME ||
-          actionResourceAndItsStackRegionComparison === TokenComparison.BOTH_UNRESOLVED;
-        const pipelineAndActionResourceStackAccountComparison = Token.compareStrings(this.env.account, actionResourceStack.account);
-        const pipelineAndActionResourceStackInSameAccount = pipelineAndActionResourceStackAccountComparison === TokenComparison.SAME ||
-          pipelineAndActionResourceStackAccountComparison === TokenComparison.BOTH_UNRESOLVED;
-        if (pipelineAndActionResourceStackInSameAccount && actionResourceInSameRegionAsItsStack) {
-          otherStack = actionResourceStack;
-        }
-      }
-    } else {
-      actionRegion = action.actionProperties.region;
+    // Do a check on whether the cross-account acccess is possible given the
+    // current pipeline configuration. This has bearing on whether
+    // `artifactBucket.grantRead(action.role)` (which is going to occur down the
+    // line) can even work.
+    if (hasAccount && !sameEnvDimension(this.env.account, actionEnv.account) && !this.crossAccountKeys) {
+      throw new Error(`Pipeline must be created with 'crossAccountKeys: true' to add cross-account action '${action.actionProperties.actionName}'. Pipeline account: '${renderEnvDimension(this.env.account)}', action account: '${renderEnvDimension(actionEnv.account)}'`);
```

Comment:
Not sure why are we checking `account` here...? There is a separate method for dealing with cross-account support resources, `getRoleFromActionPropsOrGenerateIfCrossAccount()` - I think this check belongs there. I'd rather keep this method only about cross-region concerns, without cross-account ones.